BLSQ/iaso
# IA-3728 Create change requests from differences between two pyramids (backend model)

Create change requests from differences between two pyramids (backend model).

Related JIRA tickets : [IA-3728](https://bluesquare.atlassian.net/browse/IA-3728)

## Doc

See the design specification linked in the aforementioned JIRA ticket.

## Changes

- new `DataSourceSynchronization` model and admin:
    - allows to synchronize two pyramids via "change requests"
    - this model is tightly coupled with the `diffing` module for the logic

- `diffing` module:
    - added tests (this was helpful to understand how it works)
    - fixed `Dumper.dump_as_json()` which was broken
    - fixed `Differ.diff` for groups (only groups from one side were taken into account)
    - fixed `Differ.compare_fields`
        - `if dhis2_value is None and ref_value is not None:` wasn't working when `ref_value = []` (empty list) for empty groups

## How to test

1. go to `/admin/iaso/datasourceversionssynchronization/add/`
2. configure a new "Data source synchronization" with two source versions that you want to synchronize
    - ensure those two source versions have differences
3. save
4. find the ID of the newly created "Data source synchronization"
5. open a Django shell `docker compose exec -ti iaso ./manage.py shell_plus`
6. trigger the synchronization (those steps will be handled by the future API):

    - get the "Data source synchronization" object assuming its ID is `1`:

            sync = DataSourceVersionsSynchronization.objects.get(id=1)

    - generate the diff between the two pyramids (this allows to count how many change requests there will be):

            sync.create_json_diff()

    - synchronize the two pyramids and create "change requests":

            sync.synchronize_source_versions()

7. go to `/admin/iaso/orgunitchangerequest/`
8. filter the list by your "Data source synchronization"
9. you should be able to see the newly created "change requests"

## Print screen

![admin1](https://github.com/user-attachments/assets/a0b812cf-6e52-453f-9322-c78691182d77)

![admin2](https://github.com/user-attachments/assets/d4cad823-5399-488a-a97b-ce7e49a55b22)

[IA-3728]: https://bluesquare.atlassian.net/browse/IA-3728?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
